[15721] TileGroup Compaction and GC Fixes

dqrs opened the pull request:
## Project: TileGroup Compaction, TileGroup Freeing, and Garbage Collection Fixes

Our project has 3 main aspects:

1. Enabling the Garbage Collector to free empty tile groups and reclaim their memory (Done)
2. Merging sparsely occupied tile groups to save memory (Done)
3. Fixing several important bugs in the Garbage Collector and Transaction Manager (Done)

# Status
**Completed 75% 100% and 125% goals**

# Summary

**1) TileGroup Freeing**
This PR enhances the Garbage Collector to free empty TileGroups when all of their tuple slots have been recycled. We also scanned the codebase for all points (outside of codegen) that fetch tile groups from the catalog without checking for a nullptr. TileGroups can no longer be assumed to live forever, so these checks must be done. We need to follow up with Prashanth to do a similar check in the codegen engine. It also created a new class, RecycleStack to replace  the Garbage Collector’s recycle queues. This concurrent data structure offers a non-blocking, constant-time TryPop() to retrieve a recycled tuple slot, a blocking, constant-time Push(), and a concurrent RemoveAllWithKey() function that uses hand-over-hand locking. The performance of TryPop() is very important because TryPop() lies on the critical path for DataTable::Insert(). We had to write a new data structure instead of using the existing concurrent queue because we need to be able to iterate through and remove recycled slots from TileGroups that become immutable. None of the existing concurrent structures support concurrent iteration, insertion, and removal.

**2) TileGroup Compaction**
This PR adds a class called TileGroupCompactor that performs compaction of tile groups. Compaction is triggered by the GarbageCollector, which submits a CompactTileGroup() task to the MonoQueuePool when the majority of a TileGroup is recycled garbage. The exact fraction of garbage required to trigger compaction is determined by a global setting, settings:SettingId::compaction_threshold. Compaction can also be enabled/disabled via another setting, global setting settings:SettingId::compaction_enabled. 

**3) Garbage Collection Fixes**
This PR resolves almost all of the issues identified in issue #1325. 
It includes the results of a thorough correctness audit we performed on Peleton's garbage collection system. It includes a whole new test suite for the Transaction-Level Garbage Collector and several important bug fixes to the Garbage Collector and Transaction Manager.  

**GC Fixes Summary**:

* Added 14 tests to transaction_level_gc_manager_test.cpp to handle more complex GC scenarios. 
* Currently 4 of these tests still fail for polluting indexes with old keys, but we believe this will require more significant changes at the execution layer to resolve.
*  We believe we have resolved all of the tuple-level GC bugs and most of the index bugs.
* We have disabled the 4 check that fail and will open a new issue describing those scenarios only.
* GCManager::RecycleTupleSlot allows unused ItemPointers to be returned without going through the entire Unlink and Reclaim process.
* Modified TOTransactionManager to pass tombstones created by deletes to the GCManager.
*Modified DataTable's Insert to return the ItemPointer to the GCManager in the case of a failed insert.
* Modified DataTable's InsertIntoIndexes to iterate through indexes and remove inserted keys in the event of a failure.
* Modified GCManager's Unlink function to clean indexes from garbage created by COMMIT_DELETE, COMMIT_UPDATE, and ABORT_UPDATE.

Commit:
Created GC CommitDelete Test. Made ClearGarbage public in TLGC

Commit:
Created GC CommitDelete Test. Made ClearGarbage public in TLGC

Commit:
Created GC CommitDelete Test. Made ClearGarbage public in TLGC

Commit:
Created GC CommitDelete Test. Made ClearGarbage public in TLGC

Commit:
Modified CommitDelete test so that it properly cleans up the database

Commit:
Modified CommitDelete test so that it properly cleans up the database

Commit:
Modified CommitDelete test so that it properly cleans up the database

Commit:
Modified CommitDelete test so that it properly cleans up the database

Commit:
Added 14 tests to transaction-level GC manager. Captures 4 GC failures.

Commit:
Added 14 tests to transaction-level GC manager. Captures 4 GC failures.

Commit:
Added 14 tests to transaction-level GC manager. Captures 4 GC failures.

Commit:
Added 14 tests to transaction-level GC manager. Captures 4 GC failures.

Commit:
Added test utilities, and multiple new test cases for checking
correctness of primary and secondary indexes in the garbage collector.

Commit:
Added test utilities, and multiple new test cases for checking
correctness of primary and secondary indexes in the garbage collector.

Commit:
Added test utilities, and multiple new test cases for checking
correctness of primary and secondary indexes in the garbage collector.

Commit:
Added test utilities, and multiple new test cases for checking
correctness of primary and secondary indexes in the garbage collector.

Commit:
Added more index tests. Added tests for primary key updates.

Commit:
Added more index tests. Added tests for primary key updates.

Commit:
Added more index tests. Added tests for primary key updates.

Commit:
Added more index tests. Added tests for primary key updates.

mbutrovich committed:
Added PrimaryKeyUpdateTest

mbutrovich committed:
Added PrimaryKeyUpdateTest

mbutrovich committed:
Added PrimaryKeyUpdateTest

mbutrovich committed:
Added PrimaryKeyUpdateTest

mbutrovich committed:
Refactor.

mbutrovich committed:
Refactor.

mbutrovich committed:
Refactor.

mbutrovich committed:
Refactor.

mbutrovich committed:
Fixed eror in CommitUpdateSecondaryKeyTest.

mbutrovich committed:
Fixed eror in CommitUpdateSecondaryKeyTest.

Commit:
Fixed bug where tuple slots are not reclaimed when insertions fail.
Added function RecycleUnusedTupleSlot() TransactionLevelGCManager.

Commit:
Fixed bug where tuple slots are not reclaimed when insertions fail.
Added function RecycleUnusedTupleSlot() TransactionLevelGCManager.

Commit:
Fixed bug where tuple slots are not reclaimed when insertions fail.
Added function RecycleUnusedTupleSlot() TransactionLevelGCManager.

Commit:
Fixed bug where tuple slots are not reclaimed when insertions fail.
Added function RecycleUnusedTupleSlot() TransactionLevelGCManager.

Commit:
Updated transaction manager and garbage collector to properly handle
garbage created from delete operations.

Commit:
Updated transaction manager and garbage collector to properly handle
garbage created from delete operations.

Commit:
Updated transaction manager and garbage collector to properly handle
garbage created from delete operations.

Commit:
Updated GetRecycledTupleSlot() in GC to no longer hand out slots from
immutable tile groups. Removed old/outdated immutability test that
deleted from an immutable tile group erroneously.

Commit:
Updated GetRecycledTupleSlot() in GC to no longer hand out slots from
immutable tile groups. Removed old/outdated immutability test that
deleted from an immutable tile group erroneously.

Commit:
Enhanced GC tests to check indexes in all cases, uses new test function

Commit:
Enhanced GC tests to check indexes in all cases, uses new test function

Commit:
Enhanced GC tests to check indexes in all cases, uses new test function

Commit:
Enhanced GC tests to check indexes in all cases, uses new test function

Commit:
Modified the GC so that it properly removes index entries for
deleted and updated commits and aborts. Updated the way DataTable
inserts into indexes to make insert failures more easily undone in
indexes.

Commit:
Modified the GC so that it properly removes index entries for
deleted and updated commits and aborts. Updated the way DataTable
inserts into indexes to make insert failures more easily undone in
indexes.

mbutrovich committed:
Minor refactor and comments before a formatting run.

mbutrovich committed:
Minor refactor and comments before a formatting run.

mbutrovich committed:
Minor refactor and comments before a formatting run.

mbutrovich committed:
Minor refactor and comments before a formatting run.

mbutrovich committed:
clang-format-3.6 on modified files.

mbutrovich committed:
clang-format-3.6 on modified files.

mbutrovich committed:
clang-format-3.6 on modified files.

mbutrovich committed:
clang-format-3.6 on modified files.

mbutrovich committed:
clang-format-3.6 again after rebase.

mbutrovich committed:
Disabled 4 failing tests that are not addressed in this PR and will open a new issue for those after this is merged.

mbutrovich committed:
Disabled 4 failing tests that are not addressed in this PR and will open a new issue for those after this is merged.

mbutrovich committed:
Reenabled 4 disabled tests because we still want to test recycle slots. Disabled the index checks in those tests until that issue is resolved.

mbutrovich committed:
Reenabled 4 disabled tests because we still want to test recycle slots. Disabled the index checks in those tests until that issue is resolved.

poojanilangekar committed:
Fixed unused attribute

poojanilangekar committed:
Revert "clang-format-3.6 on modified files."

This reverts commit c1bcd59

poojanilangekar committed:
Revert "clang-format-3.6 on modified files."

This reverts commit c1bcd59

poojanilangekar committed:
Revert "clang-format-3.6 again after rebase."

This reverts commit 82633f6

poojanilangekar committed:
Revert "clang-format-3.6 again after rebase."

This reverts commit 82633f6

poojanilangekar committed:
Revert some remaining format changes + disable tests

poojanilangekar committed:
Revert some remaining format changes + disable tests

Commit:
Applied all TileGroup compaction changes to the GC Fixes branch.
Added more comments in a few places.

Commit:
Applied all TileGroup compaction changes to the GC Fixes branch.
Added more comments in a few places.

mbutrovich committed:
Minor fixes after rebase.

mbutrovich committed:
More minor fixes after rebase.

mbutrovich committed:
Fixed table naming to be lowercase to fix exception in table generation.

mbutrovich committed:
Fixed errant copy-paste.

mbutrovich committed:
Removed reliance on GC Manager's tables_ structure. Go to the storage manager instead. This is for PR#2 for 15721.

Commit:
Changed tile group freeing approach. Going to use lock-free stack instead.
In progress. Untested.

Commit:
Finished initial draft of GCManager. Renamed and refactored UnlinkVersion(s)

Commit:
Created RecycleStack class. Not yet tested.

mbutrovich committed:
Bug fixes, compiles, runs transaction_level_gc_manager_tests.

mbutrovich committed:
Minor bug fixes. Passes make check. Need to write more tests to execrise new GCManager.

mbutrovich committed:
Added a way to lookup elements in LockFreeArray. Need this to do a reverse-mapping from TileGroup oid to offset. Reenabled DropTileGroupsTest. Passes make check.

Commit:
Preliminary implementation of TileGroup compaction. Needs to be tested.

mbutrovich committed:
Bug fix in compaction (removed LogicalTile), and added a new test for tile_group_compactor.

mbutrovich committed:
Made immutable_tile_group_queue a single queue instead of multiple, made access function to add to it, and made SetImmutability notify the GC.

mbutrovich committed:
Moved TileGroup compaction to a stand-alone class, separate from the GC.

suafeng commented on src/gc/tile_group_compactor.cpp:
```diff
@@ -0,0 +1,194 @@
+    }
+
+    bool success = MoveTuplesOutOfTileGroup(table, tile_group);
+
```
I find there is only one usage of function MoveTuplesOutOfTileGroup() and it is here. It seems that the tuples in the given tile group is always moved to the tile group's table. Then I think there may not be necessary to have a data table as argument in MoveTuplesOutOfTileGroup(), since we know that this table is this tile group's table.

suafeng commented on src/gc/transaction_level_gc_manager.cpp:
```diff
@@ -339,139 +278,127 @@ void TransactionLevelGCManager::AddToRecycleMap(
+  if (location.IsNull()) {
+    return INVALID_ITEMPOINTER;
+  }
 
```
I think it may be better for this function GetRecycledTupleSlot() to have a while around TryPop() instead of having while loop around GetRecycledTupleSlot, since this function name seems it guarantees to get a tuple slot.

suafeng commented on src/gc/transaction_level_gc_manager.cpp:
```diff
@@ -642,5 +569,66 @@ void TransactionLevelGCManager::UnlinkVersion(const ItemPointer location,
 }  // namespace peloton
+
+
+
```
You may not need so many blank lines here.

suafeng reviewed (commented):
Overall good job and code's quality seems to be good, I only add some comments about style and the design of functions.

ssyssy commented on src/gc/tile_group_compactor.cpp:
```diff
@@ -0,0 +1,194 @@
+    // ensure that this is the latest version
+    bool is_latest_version = tile_group_header->GetPrevItemPointer(physical_tuple_id).IsNull();
+    if (is_latest_version == false) {
+      LOG_TRACE("Skipping tuple, not latest version.");
```
Why skip the tuple here? Why not just abort the transaction?

ssyssy commented on src/gc/tile_group_compactor.cpp:
```diff
@@ -0,0 +1,194 @@
+
+    LOG_TRACE("Moving Visible Tuple id : %u, Physical Tuple id : %u ",
+              visible_tuple_id, physical_tuple_id);
+
```
Same here.

ssyssy commented on src/include/gc/transaction_level_gc_manager.h:
```diff
@@ -21,51 +21,66 @@
-    recycle_queue_map_.clear();
+    immutable_tile_group_queue_ = std::make_shared<LockFreeQueue<oid_t>>(INITIAL_TG_QUEUE_LENGTH);
+    recycle_stacks_ = std::make_shared<peloton::CuckooMap<
+        oid_t, std::shared_ptr<RecycleStack>>>(INITIAL_MAP_SIZE);
```
I think recycle_stack_map is better for understanding.

ssyssy commented on test/concurrency/testing_transaction_util.cpp:
```diff
@@ -212,6 +212,24 @@ storage::DataTable *TestingTransactionUtil::CreateTable(
+      "unique_btree_index", 1235, TEST_TABLE_OID, CATALOG_DATABASE_OID,
+      IndexType::BWTREE, IndexConstraintType::UNIQUE, tuple_schema,
+      key_schema, key_attrs, unique);
+
```
What is the 1235 here? Some random number?

ssyssy commented on src/include/storage/tile_group_header.h:
```diff
@@ -229,9 +230,15 @@ class TileGroupHeader : public Printable {
+  /*
+  * @brief Set's Immutable Flag to True. Only used by the Garbage Collector
+  */
+  inline bool SetImmutabilityWithoutNotifyingGC() {
```
Where would this function be called?

ssyssy reviewed (commented):
Nice job and really solid code here. I only find some small issues here. Hope that my comments can help.

mbutrovich committed:
Modifications to handle removing TileGroups: made manager.h actually remove elements instead of a sentinel value on removal, and got rid of TileGroupIterator.

mbutrovich committed:
2 small changes based on PR comments.

mbutrovich committed:
Refactor to Manager class for TileGroup dropping.

mbutrovich committed:
Fixed typo in manager.cpp.

mbutrovich committed:
Refactor and comments regarding immutability.

mbutrovich committed:
Reenabled old GC tests. Modified immutability test to be correct with recycled tombstones. It previously assumed no tombstones were recycled.

Commit:
Refactor of TransactionLevelGCManager. Added compaction queue.
Added compaction_threshold to global settings. Reverted breaking changes
to catalog::Manager. Removed recycling threshold. Reordered Running()
and ClearGarbage()

Commit:
Audited codebase to find all places where tile_groups dereferenced
without null checks. Added control flow changes when clear. Otherwise
added asserts.

bohanjason commented:
As my teammates said,  Nice Job ! !

bohanjason commented on src/codegen/inserter.cpp:
```diff
@@ -34,6 +34,8 @@ void Inserter::Init(storage::DataTable *table,
+  PELOTON_ASSERT(location_.IsNull() == false);
+
   // Get the tile offset assuming that it is in a tuple format
   auto tile_group = table_->GetTileGroupById(location_.block);
```
assert tilegroup != nullptr ?

Commit:
Created TileGroupCompactor concurrency test.
Moved helper funcs in TransactionLevelGCManagerTests to
TestingTransactionUtil. Removed recycling flag from TileGroupHeader.

Commit:
Added several TileGroupCompactor tests

mbutrovich committed:
Fixed two possible nullptr dereferences in TOTM, added some LOG_TRACE to help debugging.

Commit:
Final pass of comments. Added Setter for GC.compaction_threshold_.
Disabled compaction by default (set it to 1.0).

mbutrovich committed:
Added global setting for TileGroup freeing, default is off.

mbutrovich committed:
RecycleStack Doxygen comments.

mbutrovich committed:
TileGroupCompactor Doxygen comments.

mbutrovich committed:
TransactionLevelGCManager Doxygen comments.

mbutrovich committed:
Minor refactor and clang-tidy changes.

mbutrovich committed:
More minor refactors and changes based on PR feedback.

mbutrovich committed:
Added a boolean to disable compaction entirely.

mbutrovich committed:
Putting back the TileGroupIterator for future enhancement.

mbutrovich committed:
Fixed TRACE-level logging issues in GC manager, added some Doxygen comments to TileGroupHeader for new/modified functions.

mbutrovich committed:
Post-rebase fixes.

mbutrovich committed:
Fixed bug in manager.cpp after rebase. Removed unnecessary variables in tile_group_compactor_test, fixed bug in RetryTest.

mbutrovich committed:
clang-format-3.6 on modified lines.

mbutrovich committed:
Fix formatting issue on recycle_stack's include.

mbutrovich committed:
Fix unused variables when tests compiled in Release mode.

mbutrovich committed:
Fix printing issue in tile_group_compactor_test and reduced LOG_DEBUGs in GCManager.

mbutrovich committed:
Fix unused variables in Release mode.

mbutrovich committed:
Fix Debug printing issue that last commit introduced.

mbutrovich committed:
Added IsRunning member to MonoQueuePool so TileGroupCompactor can check if system is shutting down to avoid a race and stop trying to compact. Resolves segfault in tile_group_compactor_test.

mbutrovich committed:
Fix include issue for GCC on Travis.

mbutrovich committed:
Fix memory ordering issue for GCC on Travis.

mbutrovich committed:
Added more comments to disabled tests, fixed typo in tests.

mbutrovich committed:
New test scenario that triggers an assertion fail in TOTM (CommitInsertUpdatePrimaryKetTest).

Currently DISABLED.

mbutrovich committed:
Post-rebase fixes.

mbutrovich committed:
More post-rebase fixes. Passes make check again.

mbutrovich committed:
Fix condition in tile_group_compactor test that was too strict for some systems due to performance differences.

mbutrovich committed:
Minor refactor and comment changes.

tcm-marcel commented on src/gc/tile_group_compactor.cpp:
```diff
@@ -0,0 +1,160 @@
+    }
+
+    // check again now that we have ownsership
+    // to ensure that this is stil the latest version
```
Typo: *ownership* and *still*

tcm-marcel commented on src/gc/transaction_level_gc_manager.cpp:
```diff
@@ -153,13 +265,12 @@ int TransactionLevelGCManager::Unlink(const int &thread_id,
-      // a result, we can delete all the tuples from the indexes to which it
-      // belongs.
+      // since this txn's epochId is <= the global expired epoch id
+      // no active transactions can read the version. Asa result,
```
Typo: *As a*

tcm-marcel commented on src/include/gc/tile_group_compactor.h:
```diff
@@ -0,0 +1,77 @@
+#include "storage/storage_manager.h"
+#include "storage/tile_group.h"
+#include "storage/tuple.h"
+#include "threadpool/mono_queue_pool.h"
```
Can you move all includes that are not needed here to the implementation file?

tcm-marcel commented on test/concurrency/testing_transaction_util.cpp:
```diff
@@ -456,5 +474,83 @@ bool TestingTransactionUtil::ExecuteScan(
+
+ResultType TestingTransactionUtil::UpdateTuple(storage::DataTable *table,
+                                               const int key) {
+  srand(15721);
```
Is it ok to have non-deterministic stuff in the test cases? Shouldn't they be reproducible at any chance?

tcm-marcel reviewed (dismissed):
I didn't dig deep into this, since I am not very familiar with the Transaction and Garbage Collector code. I can go over it in more detail if requested.

mbutrovich committed:
Fix tile_group_compactor not recording ReadOwn behavior.

poojanilangekar commented on src/codegen/deleter.cpp:
```diff
@@ -38,6 +38,7 @@ void Deleter::Delete(uint32_t tile_group_id, uint32_t tuple_offset) {
 
   auto *txn = executor_context_->GetTransaction();
   auto tile_group = table_->GetTileGroupById(tile_group_id);
+  PELOTON_ASSERT(tile_group != nullptr);
```
Why was this line added? Is this necessary? Or was it for debugging? 
Can it cause undefined behavior in Release mode?

mbutrovich replied:
GetTileGroupById can now return a nullptr, this is just a check that we don't hit that during make check and other Debug builds. That's what most of the ASSERTs added to the execution layer are looking for.

If you're suggesting that we make it impossible to return a nullptr to the execution level, then we need to revisit how we access TileGroups entirely and go only through an iterator that will always be valid.

poojanilangekar replied:
I don't think you need to touch the execution layer. In a correct implementation, the execution layer should be completely oblivious to the GC. 

_**GetTileGroupById can now return a nullptr if the TileGroup has been dropped.**_ -> That is my biggest concern. So this thing is checked in debug which is great. But what happens in release mode? If this function gets a nullptr, it hits a segfault and peloton crashes? That's not expected behavior.

Like I said, in my overall comment, the whole point of the epoch based garbage collection is that the `TileGroup` references are removed only when you are certain that no other transaction is accessing it.  This is very similar to GC in the SkipListIndex.  The storage may get these `nullptr` references, but the execution layer should never get such a reference. 

I would suggest that we can either do a hangouts call over the weekend. Or you talk to @apavlo about this.

poojanilangekar commented on src/codegen/inserter.cpp:
```diff
@@ -34,6 +34,8 @@ void Inserter::Init(storage::DataTable *table,
 char *Inserter::AllocateTupleStorage() {
   location_ = table_->GetEmptyTupleSlot(nullptr);
 
+  PELOTON_ASSERT(location_.IsNull() == false);
```
Same question as above. Why was this added? Ideally, I believe these checks should be handled at a higher layer, probably the `DataTable`.

poojanilangekar commented on src/codegen/inserter.cpp:
```diff
@@ -54,8 +56,11 @@ void Inserter::Insert() {
-  ContainerTuple<storage::TileGroup> tuple(
-      table_->GetTileGroupById(location_.block).get(), location_.offset);
+  auto tile_group = table_->GetTileGroupById(location_.block).get();
+  PELOTON_ASSERT(tile_group != nullptr);
```
Same question. Why was this added?

poojanilangekar commented on src/common/container/cuckoo_map.cpp:
```diff
@@ -18,6 +18,9 @@
 #include "common/item_pointer.h"
 #include "common/logger.h"
 #include "common/macros.h"
+#include "common/container/lock_free_queue.h"
```
I don't think you need to include this file and `storage/data_table.h`.

poojanilangekar commented on src/common/container/lock_free_array.cpp:
```diff
@@ -53,6 +53,7 @@ template <typename ValueType>
 void LOCK_FREE_ARRAY_TYPE::Erase(const std::size_t &offset,
                                  const ValueType &invalid_value) {
   LOG_TRACE("Erase at %lu", offset);
+  PELOTON_ASSERT(lock_free_array.size() > offset);
```
Again, was this for debugging? Or have you introduced some new logic that could cause this condition?

mbutrovich replied:
No new logic. .at() will throw an exception that we never catch if it's out of bounds. Seems more correct to me to have a PELOTON_ASSERT first. Also, see Find() and Update() which already have that logic. Erase() not having it just looks like an oversight.

poojanilangekar commented on src/common/container/lock_free_array.cpp:
```diff
@@ -107,6 +108,20 @@ bool LOCK_FREE_ARRAY_TYPE::Contains(const ValueType &value) const {
   return exists;
 }
 
+template <typename ValueType>
```
Two main concerns: 
1. What happens if another thread modifies the value between line 115 and 117?
2. This seems inefficient. What is the use case? Would using a `CuckooHash` be more efficient?

poojanilangekar commented on src/common/init.cpp:
```diff
@@ -52,7 +52,7 @@ void PelotonInit::Initialize() {
 
   int parallelism = (CONNECTION_THREAD_COUNT + 3) / 4;
-  storage::DataTable::SetActiveTileGroupCount(parallelism);
+  storage::DataTable::SetDefaultActiveTileGroupCount(parallelism);
```
Nice job!

poojanilangekar commented on src/concurrency/timestamp_ordering_transaction_manager.cpp:
```diff
@@ -580,6 +580,9 @@ void TimestampOrderingTransactionManager::PerformDelete(
 }
 
+// Performs Delete on a tuple that was created by the current transaction, and
+// never
```
Combine line 584 and 585.

poojanilangekar reviewed (changes requested):
Hey @mbutrovich! I am a little confused/concerned about your changes in the execution layer. 

Ideally, the GC tasks should be transparent to the execution layer. The storage layer might access a nullptr because it is at the physical level. But the execution layer should never see the effects of GC. The whole point of epoch based GC is that it only removes those blocks/tiles/data stores that aren't visible to the executor.  (Think of something along the lines of GC in the SkipList index, you never checked the pointers during traversal right?)

There are a bunch of changes in the `updater`, `inserter`, `deleter`, `hybrid_scan_executor`, `index_scan_executor` and `seq_scan_executor` which
shouldn't be there. 

I am yet to review the rest of the code but I think we should discuss this before the rest.
